crowbar_framework/app/models/service_object.rb

```diff
@@ -1418,6 +1440,48 @@ def apply_role(role, inst, in_queue)
     [200, {}]
   end
 
+  def delete_remove_role_from_runlist(nodes_with_roles_to_remove)
+    databag = Chef::DataBag.load "crowbar/bc-template-#{@bc_name}"
+    nodes_with_roles_to_remove.each do |node_name, roles_to_remove|
+      node = NodeObject.find_node_by_name(node_name)
+      next if node.nil?
+
+      # delete #{role_name}_remove role from runlist as it is not needed anymore
+      roles_to_remove.each do |role_to_remove|
+        next unless node.role? role_to_remove
+        @logger.debug("Removing temporary role #{role_to_remove} from #{node.name}")
+        node.delete_from_run_list role_to_remove
+        save_node_state(node, __method__)
+
+        # delete _remove roles from databag
+        databag["deployment"][@bc_name]["elements"]["to_remove"][node_name].delete(role_to_remove)
+        databag.save
+      end
+
+      # delete nodes from databag with no more _remove roles
+      if databag["deployment"][@bc_name]["elements"]["to_remove"][node_name].empty?
+        databag["deployment"][@bc_name]["elements"]["to_remove"].delete(node_name)
+        databag.save
+      end
+    end
+
+    # delete remove intention from databag, as it has been done successfully
+    if databag["deployment"][@bc_name]["elements"]["to_remove"].empty?
+      databag["deployment"][@bc_name]["elements"].delete("to_remove")
+      databag.save
+    end
+  end
+
+  def save_node_state(node, origin)
```

Comment:
I would prefer `#save_node_state` to automatically calculate the caller's origin itself by introspecting the stack trace.